### ISSUE TITLE
express-jwt and express-graphql: error TS2339: Property 'user' does not exist on type 'Request'

### ISSUE THREAD

**Issue by francoisromain** [auth0/express-jwt#215]
I try to make express-jwt and graphql work together in typescript. 

I [asked about it on stackoverflow](https://stackoverflow.com/questions/54462222/express-jwt-and-express-graphql-error-ts2339-property-user-does-not-exist-on) but had no answer. So I repost here. 

```
import * as express from 'express'
import * as expressGraphql from 'express-graphql'
import * as expressJwt from 'express-jwt'

import schema from './api/schemas'
import rootValue from './api/resolvers'

const app = express()

app.use(
  expressJwt({
    credentialsRequired: false,
    secret: process.env.JWT_SECRET
  })
)

app.use(
  '/',
  expressGraphql((req, res, graphQLParams) => ({
    schema,
    rootValue,
    context: {
      user: req.user
    }
  }))
)
```

I imported the relative typings `@types/express`, `@types/express-graphql` and `@types/express-jwt`. 

There is a typescript error: 
```
error TS2339: Property 'user' does not exist on type 'Request'
```

`user` is added on the request object by `express-jwt`. 

How can I fix that?

**Comment by jellz** [comment]
I had the same problem. I fixed it by making an interface (I called mine `AuthRequest`) that extends Express' `Request` with one property:
```js
user?: string;
```

Then turn `req` into an `AuthRequest`, like so:
L19: ```js
expressGraphql((req: AuthRequest, res, graphQLParams) => ({```

**Comment by francoisromain** [comment]
@jellz Thank you. could you paste the full code for your interface please? (I am very new to typescript)

**Comment by jellz** [comment]
I'm fairly new to TypeScript too, that's why I didn't post the code. Your code is similar to mine, but we were using it for different things (I wasn't passing it through GraphQL). For your use case, that interface should work. :smiley:

**Comment by Laubeee** [comment]
What is the desired solution for this issue? Include such an Interface directly in `@types/express-jwt`?

**Comment by Beng89** [comment]
Thank to declaration merging, you can add the user property directly to express's request model like so:

```typescript
declare global {
  interface ParsedToken {
    iss: string
    sub: string
    aud: string | string[]
    iat: number
    exp: number
    azp: string
    scope: string
  }

  namespace Express {
    interface Request {
      user?: ParsedToken
    }
  }
}
```

Declaring the token interface as global allows you to add additional properties to it as well:
```typescript
declare global {
  interface ParsedToken {
    additionalProp: string
  }
}
```

You'll need to make sure to reference the typescript file that you put these declarations in though.

**Comment by download13** [comment]
> What is the desired solution for this issue? Include such an Interface directly in `@types/express-jwt`?

Ideally it would be part of the `RequestHandler` type so that any handlers downstream of the use call would know there's a user property. The only issue might be the `requestProperty` which could bring in a value at runtime. If it's a string literal typescript might be able to extract it and use it as part of a type though.

**Comment by JipSterk** [comment]
this is now [implemented](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51314). Ci should publish a new version of `6.0.x` with the new type definitions

**Comment by benmj87** [comment]
I still seem to be getting this issue due to the auth property not being optional. Am I missing something? (I have a full sample if  needed)

```typescript
interface CustomRequest extends express.Request {
  auth?: jwt.JwtPayload
};

routes.get('/thisfails', auth, (req: ExpressJwtRequest, res: Response) => {
  console.log(req.auth);
  res.send('Express + TypeScript Server');
});

routes.get('/thisworks', auth, (req: CustomRequest, res: Response) => {
  console.log(req.auth);
  res.send('Express + TypeScript Server');
});
```

**Comment by jfromaniello** [comment]
I think you are right, if you combine with `credentialsRequired: false` then `req.auth` might be undefined. I just published the fix in `v7.5.1`.

Let me know if we can close this issue 🙏🏽

**Comment by jfromaniello** [comment]
We will export a new type called `ExpressJwtRequestUnrequired`. Please use this type with `credentialsRequired: false`

Fixed in `v7.5.2`

**Comment by benmj87** [comment]
> We will export a new type called `ExpressJwtRequestUnrequired`. Please use this type with `credentialsRequired: false`
> 
> Fixed in `v7.5.2`

That's great, thanks 👍
